Use the LLVM_HOST_TRIPLE as the default for MainUtils if the triple exists

sstamenova opened the pull request:
This allows every executable that links against MainUtils (such as onnx-mlir, tests, etc.) to default to the host triple. That way if the build of llvm used a different default triple (that is often not supported by onnx mlir), it would not be the default for onnx mlir and the tests (allowing the tests to pass on the platform where onnx mlir was built). If -mtriple is specified, it still takes precedence to the default.

sstamenova commented:
This addresses #785. One thing that occurred to me was that we could add a cmake variable to control this behavior - something like `ONNX_MLIR_USE_HOST_TRIPLE` or `ONNX_MLIR_USE_DEFAULT_TRIPLE` or `ONNX_MLIR_USE_TRIPLE=host` in order to specify whether the host triple should be used as the default or the llvm default triple should be used for the default or neither.

AlexandreEichenberger commented on src/MainUtils.cpp:
```diff
@@ -288,9 +288,11 @@ string getTargetCpuOption() {
   if (mtriple != "")
     targetOptions = "--mtriple=" + mtriple;
-  // Comand cannot tolerate extra spaces. Add only when needed.
+  if (mtriple == "" && kHostTriple != "")
```
NIT:
Isn't that just a 'else if (kHostTriple != "")`

sstamenova replied:
Oh, yeah. I had somewhat more complicated logic when I was originally working on options for the change and missed this during the clean up.

caoimhinuibrian commented:
@jenkins-droid test this please

caoimhinuibrian commented:
@AlexandreEichenberger looks like a test failure (wrong result) on Z ...

AlexandreEichenberger commented:
Rebuilding, and if failing again on z390, will try to see if I can reproduce the error locally.

AlexandreEichenberger commented:
@sstamenova:  in the meanwhile, I wonder if you could not add something in the readme to indicate how to set the triple.

Also, why adding this stuff for the triple and not also for them cpu? In my experience, they are often both needed to compile for the right architecture. Could you look into that as well? That would be great.

sstamenova commented:
@AlexandreEichenberger : Do you mean adding something about `-mtriple` to the Readme? Or about the host triple that automatically gets set up now?

The triple contains the target, cpu and operating system, so it is very descriptive of what to compile for. I'm not sure what other CPU settings you might be thinking of, can you elaborate?

AlexandreEichenberger commented:
I meant as a way to help the user: as of now, we have several ways to pass it:
* directly to onnx-mlir using the -mtriple
* your new way, which is either: LLVM_HOST_TRIPLE or ONNX_MLIR_HOST_TRIPLE
* using TEST_MTRIPLE for the test.py...
maybe this is just a few too many ways... I can change the test.py (or feel free to do it) to also use the ONNX_MLIR_HOST_TRIPLE.

sstamenova commented:
> I meant as a way to help the user: as of now, we have several ways to pass it:
> 
> * directly to onnx-mlir using the -mtriple
> * your new way, which is either: LLVM_HOST_TRIPLE or ONNX_MLIR_HOST_TRIPLE
> * using TEST_MTRIPLE for the test.py...
>   maybe this is just a few too many ways... I can change the test.py (or feel free to do it) to also use the ONNX_MLIR_HOST_TRIPLE.

Ok, that makes sense. I'll add some information to the readme and I'll see about test.py.

I was looking at the options for onnx-mlir and I saw that there is an `mcpu` option as well, but not `march`. `mtriple` specifies both `mcpu` and `march` for clang/opt, but sometimes `mcpu` and `march` are used instead. I couldn't find any place where `mcpu` is specifically needed (perhaps some of the tests?), but it seems to me that maybe we should remove `mcpu` entirely OR add `march` as well. What do you think?

AlexandreEichenberger commented on test/backend/test.py:
```diff
@@ -60,10 +56,6 @@
     result_dir = tempdir.name+"/"
 print("Test info:")
 print("  temporary results are in dir:"+result_dir)
-if args.mcpu:
```
@sstamenova please remove cpu only if you also added an env var to set the cpu

AlexandreEichenberger commented on test/backend/test.py:
```diff
@@ -905,10 +897,6 @@ def prepare(cls, model, device='CPU', **kwargs):
         print(name)
         # Command
         command_list = [TEST_DRIVER]
-        if args.mcpu:
```
same here.

AlexandreEichenberger commented on MLIR.cmake:
```diff
@@ -34,6 +34,29 @@ message(STATUS "BUILD_SHARED_LIBS       : " ${BUILD_SHARED_LIBS})
+# If LLVM_HOST_TRIPLE is not available, try to include GetHostTriple.cmake to get the triple.
+# This file is available when MLIR_DIR points to a build (as opposed to an install) directory.
+# When all else fails, default to an empty string which is the old default behavior of onnx-mlir.
+if (DEFINED LLVM_HOST_TRIPLE)
```
In your comment below:

> \>**ONNX_MLIR_DEFAULT_TRIPLE**:STRING
   The default triple for which onnx-mlir will generate code. 

But this does not appear to be the case here. `ONNX_MLIR_HOST_TRIPLE` is reset here regardless of if it was defined before or not. 

My understanding is that this entire section here should be predicated on `ONNX_MLIR_HOST_TRIPLE` being undefined.

sstamenova replied:
`ONNX_MLIR_HOST_TRIPLE` is calculated each time cmake is invoked, however, `ONNX_MLIR_DEFAULT_TRIPLE` is not reset if it was defined by the user since it's a cache variable. So if I called cmake with: `-DONNX_MLIR_DEFAULT_TRIPLE=my-triple`, `ONNX_MLIR_DEFAULT_TRIPLE` will have the value of `my-triple` even after `set` is called, because the value I gave it was cached, so `set` will not overwrite it.

AlexandreEichenberger reviewed (commented):
@sstamenova This patch seems to "force" the mtriple on Z Linux. At the very least, we should be able to override it in case we want to do some cross compilation. I know of folks needing this. I added a comment above on how to fix this.

Also, in light of the fact that we should be able to change it, the new env var should be named `ONNX_MLIR_TRIPLE` as it may not necessarily be the one of the host.

I think that would still serve your purpose that it be automatically asserted if ONNX_MLIR_HOST_TRIPLE is not asserted to begin with.

Asserting the s390x-ibm-linux caused an error for one test. Will open an issue to initiate the search for a fix to that bug.

AlexandreEichenberger commented:
#796 was created to follow up on the s390x issue uncovered by this PR.

sstamenova commented:
@AlexandreEichenberger : It sounds like we want to keep TEST_MTRIPLE and TEST_MCPU for testing purposes, so I reverted the changes to test.py.

I changed the name of the new cmake variable to `ONNX_MLIR_DEFAULT_TRIPLE` since it will act as the default for `onnx-mlir` unless overwritten with the `-mtriple` option. `ONNX_MLIR_DEFAULT_TRIPLE` is a pure cmake variable in line with how llvm uses cmake variables, so it does not read its value from the environment, should it also check if it's set in the environment?

caoimhinuibrian commented:
@jenkins-droid test this please

sstamenova commented:
> #796 was created to follow up on the s390x issue uncovered by this PR.

@AlexandreEichenberger : Do we want to wait until #796 is fixed to make the triple change? Or add a workaround in the tests, so that this particular failure is hidden again?

I'd like to wait to commit the change until we update llvm to a newer commit from main because my change to export the host triple is now in main llvm and I will be able to remove the workaround in mlir.cmake. Hopefully that won't be too long.

sstamenova commented:
@AlexandreEichenberger : I made the update to this change now that we have access to the triple from llvm. This should be ready to go in unless the failure from #796 still exists.

caoimhinuibrian commented:
@jenkins-droid test this please

caoimhinuibrian reviewed (approved):
LGTM
